homeassistant/components/garmin_connect/__init__.py

```diff
@@ -86,23 +86,61 @@ def __init__(self, hass, client):
         """Initialize."""
         self.client = client
         self.data = None
+        self._activity_data = None
+        self._body_composition_data = None
 
     @Throttle(MIN_SCAN_INTERVAL)
     async def async_update(self):
         """Update data via library."""
         today = date.today()
 
         try:
-            self.data = self.client.get_stats(today.isoformat())
+            self._activity_data = self.client.get_stats(today.isoformat())
         except (
             GarminConnectAuthenticationError,
             GarminConnectTooManyRequestsError,
         ) as err:
-            _LOGGER.error("Error occurred during Garmin Connect get stats: %s", err)
+            _LOGGER.error(
+                "Error occurred during Garmin Connect get activity data: %s", err
+            )
             return
         except (GarminConnectConnectionError) as err:
-            _LOGGER.error("Error occurred during Garmin Connect get stats: %s", err)
+            _LOGGER.error(
+                "Error occurred during Garmin Connect get activity data: %s", err
+            )
             return
         except Exception:  # pylint: disable=broad-except
-            _LOGGER.error("Unknown error occurred during Garmin Connect get stats")
+            _LOGGER.error(
+                "Unknown error occurred during Garmin Connect get activity data"
+            )
             return
+
+        try:
+            self._body_composition_data = self.client.get_body_composition(
+                today.isoformat()
+            )["totalAverage"]
```

Comment:
Can this result in a KeyError?

Reply:
Even if people don't fill in weight data or don't have a Garmin scale it will return the data with None values, and will be catched by exception type Unknown already present. Nevertheless I added a KeyError exception check in last commit.